Mathlib/Analysis/Meromorphic/NormalFormAt.lean

```diff
@@ -0,0 +1,243 @@
+/-
+Copyright (c) 2025 Stefan Kebekus. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stefan Kebekus
+-/
+import Mathlib.Analysis.Meromorphic.Order
+
+/-!
+# Normal form of meromorphic functions and continuous extension
+
+If a function `f` is meromorphic on `U` and if `g` differs from `f` only along a
+set that is codiscrete within `U`, then `g` is likewise meromorphic. The set of
+meromorphic functions is therefore huge, and `=ᶠ[codiscreteWithinU]` defines an
+equivalence relation.
+
+This file implements continuous extension to provide an API that allows picking
+the 'unique best' representative of any given equivalence class, where 'best'
+means that the representative can locally near any point `x` be written 'in
+normal form', as `f =ᶠ[𝓝 x] fun z ↦ (z - x) ^ n • g` where `g` is analytic and
+does not vanish at `x`.
+
+TODO: Establish further properties of meromorphic functions in normal form, such
+as a local identity theorem. Establish the analogous notion `MeromorphicNFOn`.
+-/
+
+open Topology
+
+variable
+  {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {f : 𝕜 → E}
+  {x : 𝕜}
+
+/-!
+# Normal form of meromorphic functions at a given point
+
+## Definition and characterizations
+-/
+
+/-- A function is 'meromorphic in normal form' at `x` if it vanishes around `x`
+or if it can locally be written as `fun z ↦ (z - x) ^ n • g` where `g` is
+analytic and does not vanish at `x`. -/
+def MeromorphicNFAt (f : 𝕜 → E) (x : 𝕜) :=
+  (f =ᶠ[𝓝 x] 0) ∨ (∃ (n : ℤ), ∃ g : 𝕜 → E, (AnalyticAt 𝕜 g x) ∧ (g x ≠ 0) ∧
+    (f =ᶠ[𝓝 x] (· - x) ^ n • g ))
+
+/-- Reformulation of the definition for convenience -/
+theorem meromorphicNFAt_def :
+    MeromorphicNFAt f x ↔ (f =ᶠ[𝓝 x] 0) ∨
+    (∃ (n : ℤ), ∃ g : 𝕜 → E, (AnalyticAt 𝕜 g x) ∧ (g x ≠ 0) ∧ (f =ᶠ[𝓝 x] (· - x) ^ n • g)) :=
+  .rfl
+
+/-- A meromorphic function has normal form at `x` iff it is either analytic
+there, or if it has a pole at `x` and takes the default value `0`. -/
+theorem MeromorphicAt.meromorphicNFAt_iff (hf : MeromorphicAt f x) :
+    MeromorphicNFAt f x ↔ (AnalyticAt 𝕜 f x) ∨ (hf.order < 0 ∧ f x = 0) := by
```

Comment:
Isn't this true as
```suggestion
theorem MeromorphicAt.meromorphicNFAt_iff :
    MeromorphicNFAt f x ↔ MeromorphicAt f x ∧ (AnalyticAt 𝕜 f x ∨ hf.order < 0 ∧ f x = 0) := by
```

Reply:
This is probably true, and might well be useful … but I fear the compiler will reject your formulation because `hf.order` is not defined. There must be a way to fix that, but I cannot formulate something that compiles. Please advise.

Reply:
Here it is:
```suggestion
theorem MeromorphicAt.meromorphicNFAt_iff :
    MeromorphicNFAt f x ↔ AnalyticAt 𝕜 f x ∨ ∃ hf : MeromorphicAt f x, hf.order < 0 ∧ f x = 0 := by
```
but really `MeromorphicAt.order` should take `f` as an argument and return a junk value if `f` isn't meromorphic

Reply:
Roger. I implemented your suggestion. I tend to not touch `MeromorphicAt.order` unless this has been discussed on Zulip with the people who introduced it…